### PR TITLE
Uar 1614 ole bo new nocs

### PR DESCRIPTION
**Pull request by mattch1**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-1614

### Change description

Added the new noc fields:

"non_legal_firm_control_nature_of_control_types";
"trust_control_nature_of_control_types";
"owner_of_land_person_nature_of_control_jurisdictions";
"owner_of_land_other_entity_nature_of_control_jurisdictions";

along with validation and schema updates.

### Work checklist

- [x] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Comment by mattch1** [comment]
The DtoModelChange test fails with this message:

Unexpected NULL value for field 'nonLegalFirmControlNatureOfControlTypes'. This indicates that the DTO/DAO structure has been updated without the new fields being added to the latest version of the test JSON file or DTO/DAO generated mapper classes are not up to date

The schema, dto, dao, generated mappings and test data are all up to date so whatever the test believes to be missing is very well hidden.

**Comment by mattch1** [comment]
> The DtoModelChange test fails with this message:
> 
> Unexpected NULL value for field 'nonLegalFirmControlNatureOfControlTypes'. This indicates that the DTO/DAO structure has been updated without the new fields being added to the latest version of the test JSON file or DTO/DAO generated mapper classes are not up to date
> 
> The schema, dto, dao, generated mappings and test data are all up to date so whatever the test believes to be missing is very well hidden.

> Turns out there was a second corporate bo in the test data, updating that seems to have fixed the test. I guess one corporate bo is never enough!

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerCorporateValidatorTest.java [hunk 1]
```diff
@@ -251,11 +259,57 @@ void testErrorReportedWhenNoNatureOfControlValuesAreAllNull() {
...
+                BENEFICIAL_OWNERS_CORPORATE_FIELD,
+                BeneficialOwnerCorporateValidator.NATURE_OF_CONTROL_FIELDS);
+        String validationMessage = String.format(ValidationMessages.NOT_EMPTY_ERROR_MESSAGE, qualifiedFieldName);
+
+        assertError(BeneficialOwnerCorporateValidator.NATURE_OF_CONTROL_FIELDS, validationMessage, errors);
+    }
+
+    // Added for UAR-1597 can remove test when FEATURE_FLAG_ENABLE_PROPERTY_OR_LAND_OWNER_NOC is removed
+    @Test
+    void testErrorReportedWhenNoNatureOfControlValuesAreAllEmptyWhenNewNocsFlagIsTrue() {
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerCorporateValidatorTest.java:304]: similar for the 'No' in the title on this one

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerCorporateValidatorTest.java [hunk 2]
```diff
@@ -265,14 +319,124 @@ void testErrorReportedWhenNoNatureOfControlValuesAreAllEmpty() {
...
+
+    /*
+      When FEATURE_FLAG_ENABLE_PROPERTY_OR_LAND_OWNER_NOC_30082024 is removed choose this or one of the three below
+      do not need to keep all 4
+      Changed for UAR-1597 can keep when FEATURE_FLAG_ENABLE_PROPERTY_OR_LAND_OWNER_NOC_30082024 is removed
+      but will need to remove the word "Existing" end flag suffix and setFeatureFlag
+   */
     @Test
-    void testNoErrorReportedWhenSomeNatureOfControlValuesAreAllEmptyAndNull() {
+    void testNoErrorReportedWhenExistingNocPresentButOtherNatureOfControlValuesAreAllEmptyAndNullWhenNewNocsFlagIsTrue() {
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerCorporateValidatorTest.java:353]: in the test descriptions i would avoid using 'existing' and 'new' as in time they will all be existing and no longer new.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerCorporateDto.java [hunk 3]
```diff
@@ -68,6 +73,18 @@ public class BeneficialOwnerCorporateDto {
...
 
+    @JsonProperty(NON_LEGAL_FIRM_CONTROL_NATURE_OF_CONTROL_TYPES_FIELD)
+    private List<NatureOfControlType> nonLegalFirmControlNatureOfControlTypes;
+
+    @JsonProperty(TRUST_CONTROL_NATURE_OF_CONTROL_TYPES_FIELD)
+    private List<NatureOfControlType> trustControlNatureOfControlTypes;
+
+    @JsonProperty(OWNER_OF_LAND_PERSON_NATURE_OF_CONTROL_JURISDICTIONS_FIELD)
+    private List<NatureOfControlJurisdictionType> ownerOfLandPersonNatureOfControlJurisdictions;
+
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerCorporateDto.java:84]: Minor but one of these blank lines could be deleted.

**Review by mwestacott — COMMENTED** [review]
Will need to test these changes with the API SDK Node changes, once I've got those done.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerCorporateValidatorTest.java (hunk 2 above)
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerCorporateValidatorTest.java:353]: I agree but once the flag is removed all these tests will change anyway including the wording; that will be the time to update this.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerCorporateDto.java (hunk 3 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerCorporateDto.java:84]: I went to look and they all seem to be evenly spaced.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerCorporateValidatorTest.java [hunk 4]
```diff
@@ -251,11 +259,57 @@ void testErrorReportedWhenNoNatureOfControlValuesAreAllNull() {
...
+                BENEFICIAL_OWNERS_CORPORATE_FIELD,
+                BeneficialOwnerCorporateValidator.NATURE_OF_CONTROL_FIELDS);
+        String validationMessage = String.format(ValidationMessages.NOT_EMPTY_ERROR_MESSAGE, qualifiedFieldName);
+
+        assertError(BeneficialOwnerCorporateValidator.NATURE_OF_CONTROL_FIELDS, validationMessage, errors);
+    }
+
+    // Changed for UAR-1597 can remove test when FEATURE_FLAG_ENABLE_PROPERTY_OR_LAND_OWNER_NOC is removed
+    @Test
+    void testErrorReportedWhenNoNatureOfControlValuesAreAllEmptyWhenNewNocsFlagIsFalse() {
```
[the remark this answers is not in the document]
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerCorporateValidatorTest.java]: Removed all instances including those in the GPA tests.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerCorporateValidatorTest.java (hunk 1 above)
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerCorporateValidatorTest.java]: Removed all instances including those in the GPA tests.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by markpit** [comment]
LGTM

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerCorporateDto.java (hunk 3 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dto/BeneficialOwnerCorporateDto.java:84]: Ah right. My mistake in reading the GitHub diffs. Sorry.

**Review by mwestacott — APPROVED** [review]
Tests with SDK branch passed (a few days ago now) so approving. SDK changes merged and OE web updated so suggest one final test with this branch before merging.

**Comment by mattch1** [comment]
Also tested locally and input for the new nocs are present in the mongo database.

**Pull request merged**